content/blog/eng/from-0-to-infinite-the-final-keyword.md

```diff
@@ -0,0 +1,338 @@
+---
+authors: ["bruce"]
+date: "2018-12-04"
+draft: true
+share: false
+categories: [English, php, oop]
+title: "From zero to infinite: the Final keyword"
+type: "post"
+languageCode: "en-EN"
+toc: true
+---
+
+# Google from zero to infinite
+
+Sometimes I do a full immersion on topics of my interest, generally related to programming, topics on testing, good design, etc. 
+I apply a technique invented by me named: **"Google from zero to infinite"**.
+
+Basically I use one or more keywords in Google, then I literally follow all the links of all the pages. It is a very long and laborious activity. But most of the time I find very interesting documents, otherwise difficult to reach. This time I decided to use the keyword final of PHP.
+
+# The concept behind final keyword
+ 
+The keyword final was introduced in version 5 of php. 
+
+The manual says: 
+
+> Final Keyword. PHP 5 introduces the final keyword, which prevents child classes from overriding a method by prefixing the definition with final. 
+> If the class itself is being defined final then it cannot be extended. Note: Properties cannot be declared final, only classes and methods may be declared as final.
+
+Java has the same keyword, c# too but it is called "sealed" instead of "final", but more or less have the same behavior: to prevent a class from being inherited (in our case we will see only this type of use, for example languages like Java support further behavior but in different contexts).
+
+First of all we will try to understand what it is for, then we will try to do some additional reasoning, and finally we will talk about of comments on the net.
+
+The concept that expresses the keyword final is that a class can not be extended, the decision must be taken by the programmer in the design stage utilizing just the final keyword. 
+
+There are situations in which extension by inheritance is a good solution, others where it is not possible and others where it is wrong. The keyword final is a "tool" given to the programmer to express this decision.
+
+So when the consumer of the class will meet the keyword final, it will have to use different approach from the inheritance. And this decision is imposed from the class designer.
+
+Why someone should set such a strong constraint?. The answers are many, all aim of having a robust,  and more manageable code-base. For example to avoid the proliferation of infinite hierarchies of classes such as VCL (https://en.wikipedia.org/wiki/Visual_Component_Library), which are difficult to extend and maintain. In this case, even being a well-made library, the use of inheritance has been used a lot. 
+
+Also I use final is to reduce the API surface that needs to be covered by BC breaks avoidance in a distributed library.
+
+Further to use **final** in some cases forces adopt the interfaces, by testing for example, this is a positive side effect brought indirectly by the use of this keyword.
```

Comment:
This sentence sounds a bit strange, I try to rewrite it but I don't know if I understood it correctly. Let me know if I have changed the meaning:

Furthermore, using  **final** forces the adoption of interfaces for example, for testing. This is a positive side effect indirectly caused by the use of this keyword.